### PR TITLE
Move flood control to a trait and make the models use it

### PR DESCRIPTION
**Pull request by DaazKu**
Fixes https://github.com/vanilla/vanilla/issues/5045

This PR adds 
- CacheInterface
    - Ultimately it will inherit the SimpleCache interface from PSR
    - Add a level of abstraction
- FloodControlTrait
    - Receive a CacheInterface (which is either an adapter of UserAttributes or Gdn_Cache). The latter will be used to fix concurrency issues mentioned in vanilla/vanilla-patches/issues/65
- Some folders in library/core
    - Namely, the helpers folder. We can put utility classes in there!

### CODE REVIEW

File: applications/conversations/models/class.conversationsmodel.php [hunk 1]
```diff
@@ -13,6 +13,13 @@
  */
 abstract class ConversationsModel extends Gdn_Model {
 
+    use \Vanilla\FloodControlTrait;
+
+    /**
+     * @var \Vanilla\CacheInterface Object used to store the FloodControl data.
+     */
+    protected $floodControlStorageObject;
```
Comment (linc) [inline applications/conversations/models/class.conversationsmodel.php]: `$floodControlTracker` or `$floodgate` or even `$floodController` would be nicer. "StorageObject" makes it sound like I woke up in a Java shop.

Comment (linc) [inline applications/conversations/models/class.conversationsmodel.php]: I vote floodgate, myself.

Comment (initvector) [inline applications/conversations/models/class.conversationsmodel.php]: Second.

File: .travis.yml [hunk 2]
```diff
@@ -35,6 +36,6 @@ script:
   - tests/travis/php-lint.sh ./library
   - tests/travis/php-lint.sh ./plugins
   - tests/travis/php-lint.sh ./themes
-  - phpunit -c phpunit.xml.dist
+  - ./vendor/bin/phpunit -c phpunit.xml.dist --coverage-clover=coverage.clover
```
Comment (DaazKu) [inline .travis.yml]: Yes they are. Unless I rebase.

Comment (initvector) [inline .travis.yml]: Yeah, they're identical to what's already in core. The diff just threw me off. Nevermind!

**Comment by initvector** [comment]
Need to update the `$floodControlStorageObject` references in the discussion model and then I think this'll be ready to go.

**Pull request merged**